### ISSUE TITLE
Refactor cluster tests

### ISSUE THREAD

**Issue by rwrzesien** [golemfactory/concent#80]
Our cluster tests base is growing pretty fast and we need to take a wider look on how they are written. If we do this now, we can add next tests in more efficient way. We want to change general idea on how we want to create and use them. This will be done in two parts:

1. Create a python module which will be very generic, and able to perform request to to given cluster url and verify response. It should have following parameters:

- Golem message type.
- Message parameters (might be separate for each parameter but dict would be also nice and more generic).
- Cluster url.
- Custom settings (might be separate for each parameter but dict would be also nice and more generic).
- Sleep time after/before request call.
- Role of the request sender (`provider` or `requestor`).
- Endpoint to which request is sent (`send`, `receive` or `receive_out_of_band`).
- Custom headers (like public keys)(might be separate for each parameter but dict would be also nice and more generic).
- Expected response status code.
- Expected response message type.

Also please follow this guidelines:
- Each golem message is generated in a specific way, probably a function for each will be needed, but please keep it as generic as possible.
- If script will not have a settings passed, it should call settings endpoint for given cluster url, parse and use them.
- Some generic functions for outputing request and response are ok, you can reuse existing ones.
- It would be good to separate requests creation/call from printing, currently it is bounded tightly. 

2. Create several bash scripts which will test different scenarios. In perfect world those scripts should only make several calls to python module described above with specific parameters. 

Example: (tests https://github.com/golemfactory/concent/blob/master/docs/source/communication-examples.rst#concent-passes-computed-task-acceptance-or-rejection-to-the-provider) 

```
#!/bin/bash

python name_of_the_script.py cluster_url="dev.concent" --message=ForceReportComputedTask --task_id=1 --deadline="2017-12-1 11:00:00" --role="provider" --endpoint="send"

python name_of_the_script.py cluster_url="dev.concent" --role="requestor" --endpoint="receive"

python name_of_the_script.py cluster_url="dev.concent" --message=AckReportComputedTask --task_id=1 --deadline="2017-12-1 11:00:00" --role="requestor" --endpoint="send"

python name_of_the_script.py cluster_url="dev.concent" --role="provider" --endpoint="receive"
```

**Comment by cameel** [comment]
1) > name_of_the_script.py

    Let's call the script `concent-request.py`.

2) > Role of the request sender (provider or requestor).

    I don't think it's necessary. It does not change the request in any way. Messages from requestors and providers look the same. We just need an option to pass the private key to the script (or have it generate one).

    Note that we only need to pass the private key to the script because the public key can be recreated from the private one (golem-messages has a function that does this).

3) Let's have some positional parameters. I suggest this:

    ``` bash
    ./concent-request.py <endpoint> <message type> <cluster>
    ```

    Example:
    ``` bash
    cluster=staging.concent.golem.network

    ./concent-request.py send ForceReportComputedTask $cluster --task_id=1 --deadline="2017-12-1 11:00:00"
    ```

4) Error handling. The script should not try to be too clever with errors. It should just fail when it encounters one.

5) The script should handle `Content-Type` correctly. Always try to decode the response body as the format declared in the `Content-Type` header. If the data cannot be decoded, the header is wrong or is missing when body is non-empty, the script should report an error and print the raw response.

    See the reports for the current cluster testing scripts and avoid making similar mistakes:  #46, #72, #83.

    You can expect the following content types in typical server responses (successful ones and errors):
    - JSON (`application/json`)
    - Golem message (`application/octet-stream`)
    - HTML
    - Text
    - No content type (when body is empty)

6) The script should have at least two different verbosity modes: one with lots of output for one-off calls and one with just basic output for use in scripts.

    Errors should always be printed.

7) > Custom settings (might be separate for each parameter but dict would be also nice and more generic).

    The script should get timing constants from the `protocol-constants/` endpoint by default. User should be able to override it by telling it to import it from local settings (let's call it `--constants-from-local-settings`)

8) Concent's public key can be specified on the command line or taken from local settings.

9) The script should print Concent version reported by the server in response headers. But only in verbose mode.

    Maybe it would be useful to separate logging and normal console output. E.g. have only the important information printed normally and use Django's `LOGGING` settings to determine whether to print extra info.

**Comment by cameel** [comment]
I have extracted the second part of this task (rewriting the existing scripts) into a separate issue. I have also rewritten the example to add some missing elements.

10) We need the script to be able to compare the results it got with our expectations.

    @rwrzesien could you look at the example, think what exactly we need in this regard and add it to the issue description?

**Comment by dybi** [comment]
#169 - definition for first iteration

**Comment by rwrzesien** [comment]
@cameel 
> @rwrzesien could you look at the example, think what exactly we need in this regard and add it to the issue description?

Do you mean something like this ?

```
        self._test_response(
            response_1,
            status          = 200,
            key             = self.PROVIDER_PRIVATE_KEY,
            message_type    = message.concents.ForceReportComputedTaskResponse,
            fields          = {
                'timestamp':                                                    self._parse_iso_date_to_timestamp("2017-12-1 11:00:00"),
                'reject_report_computed_task.timestamp':                        self._parse_iso_date_to_timestamp("2017-12-1 11:00:00"),
                'reject_report_computed_task.reason':                           message.RejectReportComputedTask.REASON.TaskTimeLimitExceeded,
                'reject_report_computed_task.task_to_compute.compute_task_def': compute_task_def
            }
        )
```

I would suggest checking expected results in our integration tests, but off course think through if they are correct.

**Comment by cameel** [comment]
No, I meant the `--expect-xxx` options. The script was meant to be able to compare the stuff it gets from the server with our expectations.

Remember that it's a command-line tool. We don't want to do stuff like parsing and comparing the message from the server in bash. We want the tool to be ale to check it for us.
